### DISCUSSION TITLE
Workbox fork still needed?

### DISCUSSION THREAD

**Discussion by florentmsl** [serwist/serwist#120] [General]
With the news that a team at Google will resume maintaining Workbox, will you keep the forked Workbox core or will you reference Workbox itself?

Could anyone provide insights on the differences between the two, focusing solely on the Workbox components (excluding any Nextjs-related elements)? 

TLDR: Is it possible to just add a ref to Workbox itself instead of maintaining the Workbox fork?

**Comment by DuCanhGH** [comment]
@florentmx first, I am just quoting the comment that, in my opinion, indirectly helped resurrect the Workbox project:

> Given the radio silence over the past year, I figured it would be nice to do a bit of shameless plugging here. I'd like to invite some willing contributors to [Serwist](https://9-0-0.serwist.pages.dev) (Serwist -> SW, Service (workers), Swiss), a fork of Workbox initiated in October last year. It is not marked as a fork (doing so interferes with GitHub Search), but the commit history is kept intact. At the moment, after a lot of work has been poured into the project, the differences between Workbox and Serwist are fairly noticeable:
> 
> * Serwist is a pnpm monorepo powered by Turborepo, mostly written in TypeScript, whereas Workbox is a lerna monorepo with JS taking up around 67,3% of the codebase.
> * GenerateSW has been removed. InjectManifest and GetManifest are now the only supported modes.
> * `@serwist/sw` was completely rewritten to support a completely different use case. `@serwist/sw.installSerwist` is a replacement for GenerateSW.
> * Serwist doesn't ship a CDN build. Not sure whether it works with cdnjs, jsdelivr,... though.
> * Serwist has built-in integrations for Next.js, SvelteKit, Nuxt, Vite and plans to extend this support to Remix and Astro as well (these are forks of `@ducanh2912/next-pwa` and Vite PWA's projects).
> * Serwist is an ESM-only package, whereas Workbox is a CJS one.
> * Serwist only supports Node.js >= 18 and TypeScript >= 5.
> * Serwist ships declaration maps.
> * Serwist has implemented a few parts of the Workbox v7 milestone (concurrent precaching, renamed `RuntimeCaching.urlPattern` to `RuntimeCaching.matcher`).
> * Serwist uses Zod instead of AJV.
> * Serwist is missing a lot of tests and examples. Would love some help with this one!
> * I'm not exactly sure if Serwist's CLI is working as intended (never bothered to check...).
> * Serwist _desperately_ needs help with the docs! We brought over the docs of `workbox-background-sync`, `workbox-broadcast-update`, `workbox-expiration`, and `workbox-cacheable-response` and added the docs of `@serwist/build`, `@serwist/next`, and `@serwist/vite`, but the rest remains to be implemented. This docs is implemented in Svelte rather than Markdown ([sorry about that](https://github.com/serwist/serwist/blob/9.0.0/contributing/docs/adding-docs.md#why-not-markdown)), so I'm not sure if anyone's willing to contribute. Quite an eyesore for newcomers I'd say.
> 
> We are currently iterating on [v9](https://github.com/serwist/serwist/discussions/66). It is nearly time we wrapped up the work though. Currently forking some of Workbox's docs and adding our own pages.
> 
> Would love some help on fixing some of Workbox's opened issues that apply to Serwist as well and implementing some more of the Workbox v7 milestone!
> 
> * Make `revision` optional when `integrity` is set in precache manifest?
> * Consider adding `maxUnusedSeconds` and `updateStale` to `ExpirationPlugin` config?
> 
> ![Jeff's star](https://github.com/GoogleChrome/workbox/assets/75556609/d1f3f731-9598-4ac5-bb8f-e8a8b6d2b651)
> 
> P/s: got Jeff's star! Really proud of this one.

Within the core project, even more changes have been made since that comment as well:

- Merged all service worker packages into one single entity, namely `serwist`.
- Deprecated `PrecacheController`, `Router`, and `installSerwist` in favor of `serwist.Serwist`. This new class also doesn't have a singleton instance.
- Deprecated all legacy service worker packages, which are `@serwist/core`, `@serwist/background-sync`, `@serwist/broadcast-update`, `@serwist/cacheable-response`, `@serwist/core`, `@serwist/expiration`, `@serwist/google-analytics`, `@serwist/navigation-preload`, `@serwist/precaching`, `@serwist/range-requests`, `@serwist/routing`, `@serwist/strategies`, and `@serwist/sw`.
- Got the CLI working again.
- Added `maxAgeFrom`.
- Allow non-`Promise` return types for `SerwistPlugin` callbacks.
- Added support for a `matcher` when using `PrecacheFallbackPlugin`.
- Removed support for string handlers when using `runtimeCaching`.
- Renamed `RuntimeCaching.urlPattern` to `RuntimeCaching.matcher`.
- Made `revision` optional when using `integrity`.

Serwist did indeed derive from Workbox, but you just know that Workbox can't just go and make some of the noticeable design changes listed here. And no, there is no plan to discontinue the project even with Workbox being maintained again. Well, I hope the two projects can co-exist in harmony :) It's up to you to choose which one to use though. For Next.js, I also offer `@ducanh2912/next-pwa`, which is a `next-pwa` fork compatible with App Router and makes use of Workbox.

P/s: this sentence:

> I'm considering whether it's still beneficial to maintain this repository as a fork

reads like _you_ are the maintainer 😅

**Comment by florentmsl** [comment]
Great, thank you for this insight. So my suspicion is correct, that the workbox core changed too much in order to unify it. I‘ll stick with serwist in this case. Thanks for the amazing project

> reads like you are the maintainer 😅
you are right, I changed it to avoid confusion
